### PR TITLE
Added a gallery example for saving & loading GenericMap using asdf

### LINKED ISSUE #431: Support for saving intermediate data products
Is there a good way to save intermediate SunPy data products? The pickle module does not work properly on some SunPy objects at the moment (e.g. maps, where the header information gets lost).

As an example, I may want to process some data only once (maybe it's time consuming) and generate some products which I then want to work with or manipulate many times, over multiple Python sessions.

### LINKED ISSUE #3961: Write example showing how to save GenericMap with asdf
As far as I can see #2366 added support to save a GenericMap using asdf. It would be good to add an example to this section: https://docs.sunpy.org/en/stable/generated/gallery/index.html#saving-and-loading-data showing how to do that.

Fixing this would close #431

### PR DESCRIPTION
**Pull request by chouhanaryan**
### Description

Added a gallery example that covers saving and loading GenericMaps using asdf.

Fixes #3961 
Fixes #431

### CODE REVIEW

File: examples/saving_and_loading_data/genericmap_in_asdf.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+"""
+========================================
+Saving and loading GenericMaps with asdf
+========================================
+
+In this example we are going to look at how we can save and load a
+`~sunpy.map.GenericMap` with `asdf <https://asdf.readthedocs.io/en/latest/>`__.
+
+asdf is a modern file format designed to meet the needs of the astronomy
+community. It has deep integration with Python and SunPy and Astropy as well as
```
Comment (Cadair) [inline examples/saving_and_loading_data/genericmap_in_asdf.py:10]:
```suggestion
community. It has deep integration with Python, SunPy and Astropy as well as
```

File: examples/saving_and_loading_data/genericmap_in_asdf.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+aia_map.peek()
+
+################################################################################
+# We can now save this object to an asdf file to use later. Saving it like this
+# allows us to preserve all of the metadata of the object along with the actual
+# array data. When we load the asdf again, we get an identical
+# `~sunpy.map.sources.sdo.AIAMap` object.
+#
+# asdf files work by saving a dictionary internally, so to save the object we
+# need to put it into a dictionary. This becomes what asdf calls a `tree`.
```
Comment (Cadair) [inline examples/saving_and_loading_data/genericmap_in_asdf.py:39]:
```suggestion
# need to put it into a dictionary. This becomes what asdf calls a "tree".
```

File: examples/saving_and_loading_data/genericmap_in_asdf.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+tree = {'aia_map': aia_map}
+
+################################################################################
+# We can now write the data to an asdf file like so:
+with asdf.AsdfFile(tree) as asdf_file:
+    asdf_file.write_to("sunpy_map.asdf")
+
+################################################################################
+# This asdf file is a portable file and can be safely loaded by anyone with
+# Astropy and SunPy installed. We can reload it like so:
```
Comment (Cadair) [inline examples/saving_and_loading_data/genericmap_in_asdf.py:49]:
```suggestion
# Astropy, SunPy and asdf installed. We can reload it like so:
```

File: examples/saving_and_loading_data/genericmap_in_asdf.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
+"""
+========================================
+Saving and loading GenericMaps with asdf
+========================================
```
Comment (Cadair) [inline examples/saving_and_loading_data/genericmap_in_asdf.py:4]:
```suggestion
=======================================
Saving and loading sunpy Maps with asdf
=======================================
```

**Review by Cadair — APPROVED** [review]
Thanks for writing this, it looks great! A few minor comments :smile:

**Comment by Cadair** [comment]
This just needs a second review now :smile:

File: changelog/4494.docfix.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+Added a gallery example for saving and loading GenericMaps using asdf.
```
Comment (nabobalis) [inline changelog/4494.docfix.rst:1]:
```suggestion
Added a gallery example for saving and loading sunpy Maps using asdf.
```

**Review by nabobalis — APPROVED** [review]

**Comment by nabobalis** [comment]
Thanks again @chouhanaryan

**Pull request merged**